backend/tests/http/test_dataproviders.py

```diff
@@ -0,0 +1,144 @@
+# pylint: disable=protected-access
+""" Test for token and treasury data providers' adapters
+"""
+
+from datetime import datetime, timedelta
+from json import loads
+from json.decoder import JSONDecodeError
+
+from dateutil.tz import UTC
+from dateutil.utils import today as today_in
+from fakeredis import FakeRedis
+from httpx import AsyncClient
+from pytest import MonkeyPatch, mark, raises
+
+from backend.app.libs.coingecko import coins
+from backend.app.libs.types import Price
+
+from ..conftest import HTTPStatusError, raise_http_status_error_404
+
+# Coingecko Tests
+
+
+class MockResponse:
+    status_code = None
+    text = ""
+
+    def raise_for_status(self):
+        return None
+
+    @staticmethod
+    def json():
+        return {
+            "prices": [
+                [1590019200000, 0.9993188217729783],
+                [1590105600000, 0.9981200719764681],
+                [1590192000000, 0.9995106820722109],
+                [1590278400000, 0.9984064931995519],
+                [1590364800000, 0.9941101041414783],
+            ]
+        }
+
+
+async def return_mocked_resp(*_, **__):
+    return MockResponse()
+
+
+def mocked_datetime(offset: int = 0, use_today: bool = False):
+    _mocked_datetime = datetime(2022, 1, 1) if not use_today else today_in(UTC)
+    return _mocked_datetime + timedelta(offset)
+
+
+@mark.asyncio
+async def test_coins_success(monkeypatch: MonkeyPatch):
+    monkeypatch.setattr(AsyncClient, "get", return_mocked_resp)
+
+    resp = await coins.get_data("", mocked_datetime(), mocked_datetime(5))
+
+    assert resp == [
+        [1590019200000, 0.9993188217729783],
+        [1590105600000, 0.9981200719764681],
+        [1590192000000, 0.9995106820722109],
+        [1590278400000, 0.9984064931995519],
+        [1590364800000, 0.9941101041414783],
+    ]
+
+
+@mark.asyncio
+async def test_coins_status_err(monkeypatch: MonkeyPatch):
+    monkeypatch.setattr(AsyncClient, "get", return_mocked_resp)
+    monkeypatch.setattr(MockResponse, "raise_for_status", raise_http_status_error_404)
+
+    with raises(HTTPStatusError) as error:
+        prices = await coins._get_data("", mocked_datetime(), mocked_datetime(5))
+
+        assert not prices
+        assert "mocked http status error" in error.value
+
+
+@mark.asyncio
+async def test_coins_successful_retry(monkeypatch: MonkeyPatch):
+    monkeypatch.setattr(AsyncClient, "get", return_mocked_resp)
+    monkeypatch.setattr("backend.app.libs.coingecko.coins.sleep", lambda *_: None)
+    monkeypatch.setattr(
+        MockResponse, "text", "rate limit reached\nplease try again in a minute"
+    )
+
+    prices = await coins.get_data("", mocked_datetime(), mocked_datetime(5))
```

Comment:
I'm a tiny bit puzzled when reading this test. It doesn't really test that the retry is done, but only that when some given `text` attribute is there, the function still returns something. Said another way, the test probably too much dependent on leaky abstractions.

I'd say that's fine for now, because the test is still a bit useful. But to make the retry logic more robust if we need later, we'd have to think about how to rewire things a bit.